### PR TITLE
[InstCombine] Propagate disjoint flags during shl-binop transform

### PR DESCRIPTION
**Pull request by AZero13**
Proof: https://alive2.llvm.org/ce/z/D2jHrn

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp [hunk 1]
```diff
@@ -1259,6 +1266,50 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
       match(Op1, m_SpecificIntAllowPoison(BitWidth - 1)))
     return new ZExtInst(Builder.CreateIsNotNeg(X, "isnotneg"), Ty);
 
+  Value *Y;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1269]: Please drop changes from https://github.com/llvm/llvm-project/pull/88193

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1269]: Did

**Comment by goldsteinn** [comment]
LGTM, wait on 1 more.

**Comment by dtcxzyw** [comment]
> Proof: https://alive2.llvm.org/ce/z/7YWJ69

The proof doesn't match your implementation.

File: llvm/test/Transforms/InstCombine/shl-bo.ll [hunk 2]
```diff
@@ -294,6 +294,23 @@ define i8 @lshr_and_or(i8 %a, i8 %y)  {
   ret i8 %l
 }
 
+define i8 @lshr_and_or_disjoint(i8 %a, i8 %y)  {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/shl-bo.ll:297]: Missing test with ashr.

Please just add a new commit. Don't force push again.

**Review by dtcxzyw — APPROVED** [review]
LGTM w/ additional tests.

**Pull request merged**